atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/async/CqlClientFactory.java

```diff
@@ -0,0 +1,168 @@
+import com.datastax.driver.core.policies.DefaultRetryPolicy;
+import com.datastax.driver.core.policies.LatencyAwarePolicy;
+import com.datastax.driver.core.policies.LoadBalancingPolicy;
+import com.datastax.driver.core.policies.RoundRobinPolicy;
+import com.datastax.driver.core.policies.TokenAwarePolicy;
+import com.datastax.driver.core.policies.WhiteListPolicy;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import com.palantir.atlasdb.cassandra.CassandraKeyValueServiceConfig;
+import com.palantir.atlasdb.cassandra.CassandraServersConfigs;
+import com.palantir.conjure.java.config.ssl.SslSocketFactories;
+import com.palantir.logsafe.SafeArg;
+
+import io.netty.channel.socket.SocketChannel;
+import io.netty.handler.proxy.Socks5ProxyHandler;
+
+public final class CqlClientFactory {
+
+    private static final Logger logger = LoggerFactory.getLogger(CqlClientFactory.class);
+
+    private CqlClientFactory() {
+
+    }
+
+    public static CqlClient constructClient(CassandraKeyValueServiceConfig config, boolean initializeAsync) {
+        return config.servers().accept(new CassandraServersConfigs.Visitor<CqlClient>() {
+            @Override
+            public CqlClient visit(CassandraServersConfigs.DefaultConfig defaultConfig) {
+                return new ThrowingCqlClientImpl();
+            }
+
+            @Override
+            public CqlClient visit(CassandraServersConfigs.CqlCapableConfig cqlCapableConfig) {
+                return createClient(
+                        config,
+                        cqlCapableConfig.cqlHosts(),
+                        cqlCapableConfig.socksProxy(),
+                        initializeAsync);
+            }
+        });
+    }
+
+    private static CqlClient createClient(
+            CassandraKeyValueServiceConfig config,
+            Set<InetSocketAddress> servers,
+            Optional<SocketAddress> proxy,
+            boolean initializeAsync) {
+
+        Cluster.Builder clusterBuilder = Cluster.builder()
+                .addContactPointsWithPorts(servers)
+                .withCredentials(config.credentials().username(), config.credentials().password())
+                .withCompression(ProtocolOptions.Compression.LZ4)
+                .withLoadBalancingPolicy(loadBalancingPolicy(config, servers))
+                .withPoolingOptions(poolingOptions(config))
+                .withQueryOptions(queryOptions(config))
+                .withRetryPolicy(DefaultRetryPolicy.INSTANCE)
+                .withoutJMXReporting()
+                .withThreadingOptions(new ThreadingOptions());
+
+        proxy.ifPresent(proxyAddress -> clusterBuilder.withNettyOptions(new SocksProxyNettyOptions(proxyAddress)));
+
+        sslOptions(config).ifPresent(clusterBuilder::withSSL);
+
+        Cluster cluster = clusterBuilder.build();
+
+        ThreadFactory threadFactory = new ThreadFactoryBuilder()
+                .setNameFormat(cluster.getClusterName() + "-session" + "-%d")
```

Comment:
Internal product that creates multiple instances of TransactionManagers might have issues here, because cluster name will be the same.

Reply:
Ideally we would also provide a way for that product to provide the threadpool to use here.

Reply:
This is a temp solution to keep the PRs simple. The idea was to keep this for the current related PRs and provide a way to specify how clusters and sessions will be managed.